plugins/cpp/parser/src/cppparser.cpp

```diff
@@ -329,19 +332,142 @@ int CppParser::worker(const clang::tooling::CompileCommand& command_)
 
 CppParser::CppParser(ParserContext& ctx_) : AbstractParser(ctx_)
 {
-  (util::OdbTransaction(_ctx.db))([&, this] {
-    for (const model::BuildAction& ba : _ctx.db->query<model::BuildAction>())
-      _parsedCommandHashes.insert(util::fnvHash(ba.command));
-  });
 }
-  
+
 std::vector<std::string> CppParser::getDependentParsers() const
 {
   return std::vector<std::string>{};
 }
 
+bool CppParser::preparse()
+{
+  std::vector<model::FilePtr> filePtrs(_ctx.fileStatus.size());
+
+  std::transform(_ctx.fileStatus.begin(),
+                 _ctx.fileStatus.end(),
+                 filePtrs.begin(),
+                 [this](const auto& item)
+                 {
+                   if (item.second == IncrementalStatus::MODIFIED)
+                   {
+                     return _ctx.srcMgr.getFile(item.first);
+                   }
+                   else
+                   {
+                     return std::make_shared<model::File>();
+                   }
+                 });
+
+  try
+  {
+    (util::OdbTransaction(_ctx.db))([&]
+    {
+      for (const model::FilePtr file : filePtrs)
```

Comment:
Capturing this shared ptr via copy is way too expensive for this simple operation! The `filePtrs` lives on the local stack so it's impossible for the last reference to go out of scope and destroy these objects. And because the type of the elements is self-explanatory, and only a few lines above, you can safely use `auto` without sacrificing readability.

    for (const auto& file : filePtrs)

Reply:
Thanks for the remark on how the stack and references in C++ work, it was just a typo. Fixed in 52b7b72b31bdc585c43440a29071f0842e4570b6.